Refactor assigns clause implementation

SaswatPadhi opened the pull request:
I was planning to merge these changes as part of #6249, but given the volume of changes, I thought it would be best to create a separate PR.

Main changes are:
+ bug fixes in containment (previously called alias) and subset (previously called subset) checks
+ usused fields (e.g., `function_id`, `init_block`) are removed,
+ redundant functions (e.g. `pointer_for`) are removed,
+ const references are used throughout consistently
+ targets are tracked in an `unordered_set` as opposed to `vector`
+ introduced a "normalize" method for extracting base object of arrays

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

SaswatPadhi committed:
Refactor assigns clause implementation

+ usused fields (e.g., `function_id`, `init_block`) are removed,
+ redundant functions (e.g. `pointer_for`) are removed,
+ const references are used throughout consistently
+ targets are tracked in an `unordered_set` as opposed to `vector`
+ introduced a "normalize" method for extracting base object of arrays
+ API renamed for clarity

SaswatPadhi commented:
Marked `draft` for now.

`assigns_enforce_16` is failing but I cannot reproduce the failure locally!

feliperodri commented on src/goto-instrument/contracts/assigns.cpp:
```diff
@@ -19,129 +19,107 @@ Date: July 2021
-    "Assigns clause targets should be stored as pointer expressions.");
 }
 
-assigns_clause_targett::~assigns_clause_targett()
-{
-}
```
Why remove it? Don't we need to clean any memory for `assigns_clause_targett` objects?

SaswatPadhi replied:
Now we store `targett` (not `targett *`) in our `unordered_set`. We use `emplace` when inserting and don't seem to be making any copies elsewhere as well, so I thought this was cleaner than keeping raw pointers around.

Once the raw pointers and calls to `new` were removed, there's no leftover memory to clean.

SaswatPadhi committed:
Remove dead symbols from assignable set

SaswatPadhi commented:
> Marked `draft` for now.
> 
> `assigns_enforce_16` is failing but I cannot reproduce the failure locally!

Fixed this by removing dead variables from freely assignable set all together (although the implementation is still incorrect and, as discussed offline, @feliperodri would make a separate PR to fix that properly).

But, what's concerning here is that `goto-cc` / `goto-instrument` was generating different goto binaries on my machine vs CI.

EDIT: The remaining CI issues with Windows tests seem completely unrelated.

jimgrundy commented:
I realize I'm late here and that the coding practices for this code base are established ... but I would want to see more comments in the code. While particularly valuable for a new comer like myself, a header comment per method stating what it does, and maybe also per method argument/result, would help a great deal. I imagine one day being able to run doxygen on the code and learning something useful from it.

SaswatPadhi commented:
Will push a few more things before final review & merge, so marked as `draft`.

SaswatPadhi committed:
Fix alias checks within assigns clause

assigns_clauset was calling `compatible_expression` (not
`alias_expression`) on each target, and was thus only comparing the
object IDs (not offsets and sizes).
We fix this by removing `compatible_expression` entirely and calling
`alias_expression` for each from assigns_clauset.

feliperodri reviewed (approved):
Thank you for all the subsequent clean up and fixes!
